rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+
+### The tf.ExtensionType Protocol
+
+Classes that implement the `tf.ExtensionType` protocol can be used with
+"generic" APIs whose behavior does not depend on the value of each object (such
+as `tf.function`, `SavedModel`, and `tf.while_loop`).  In order to implement
+this protocol, a class's values must be immutable and decomposable into two
+parts:
+
+* A collection of Tensors, which encodes the value's dynamic data -- i.e., data
+  that may change for different executions of a graph.
+
+* An instance of a TypeSpec subclass, which encodes the value's static data --
+  i.e., data that is the same for all executions of a graph.  (Each extension type
+  implements its own TypeSpec subclass.)
+
+As an example, consider `tf.RaggedTensor`, which adds ragged dimensions to a
+`flat_values` `Tensor` by using `row_partition` tensors to divide it into
+sublists.  Its **_dynamic data_** consists of the `flat_values` tensor and the
+list of `row_partition` tensors (one for each ragged dimension).  Its **_static
+data_**, which consists of the dtype and static shape for `flat_values`, the
+number of ragged dimensions, and the dtype used to encode the `row_partition`
+tensors, is stored using an instance of `tf.RaggedTensorSpec`.
+
+As another example, consider a hypothetical `LabeledDigraph` class, which
+encodes a directed graph with data on both nodes and edges.  Its **_dynamic
+data_** could consist of: (1) a pair of a string-keyed dictionary of node
+tensors with shape `[num_nodes, ...]`; (2) a string-keyed dictionary of edge
+tensors with shape `[num_edges, ...]`; and (3) a pair of integer tensors
+specifying the source and destination node index for each edge.  Its **_static
+data_**, which would include information about the dtypes and static shapes of
+all the node and edge label tensors, and would be stored in a
+`LabeledDigraphSpec` class.
+
+The work of decomposing values into parts and reconstructing values from those
+parts is handled by the extension type's `TypeSpec` subclass.  Thus, the
+`tf.ExtensionType` protocol just requires that we provide a `TypeSpec` for each
+value:
+
+```python
+class ExtensionType(Protocol):
+  """Protocol for defining TensorFlow extension types.
+
+  TensorFlow extension types must be immutable, and their values must be
+  decomposable into two parts:
+
+  * A collection of Tensors, which encodes the value's dynamic data
+    (i.e., data that may change for different executions of a graph).
+  * An instance of `TypeSpec`, which encodes the value's static data
+    (i.e., data that is the same for all executions of a graph).
+
+  The `TypeSpec` is returned by `self.__tf_type_spec__()`; and the collection
+   of tensors is returned by `self.__tf_type_spec__().to_components(self)`.
+  """
+  def __tf_type_spec__(self): TypeSpec
+    """The `TypeSpec` describing the type for this value."""
+    raise NotImplementedError
+```
+
+**Note:** `tf.ExtensionType` is a Python `Protocol`, so it does not need to be
```

Comment:
Note that typing.Protocol is new in Python 3.8. I presume you'll want to preserve some degree of backwards compatibility older versions of Python? Is so, how is that handled?

Reply:
We will define an `tf.is_extension_type` method which functions appropriately in all versions of Python.